File: tools/append_event.py
#!/usr/bin/env python3
"""Append one ledger event from a Grok bot. Same contract as tools/append-event.mjs.

The cycle stays in this repo: events land in ledger/events.jsonl. Do not POST
anywhere. Lovable is a public view of site/, not part of the agent loop.
"""
from __future__ import annotations

import json
import os
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT / "tools"))
from ledger_contract import ContractError, validate as contract_validate  # noqa: E402


def load_env(path: Path) -> None:
    if not path.exists():
        return
    for line in path.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, _, val = line.partition("=")
        os.environ.setdefault(key.strip(), val.strip())


def ledger_path() -> Path:
    return Path(os.environ.get("LEDGER_PATH", ROOT / "ledger" / "events.jsonl"))


def append(event: dict) -> Path:
    path = ledger_path()
    contract_validate(event, path)
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a") as f:
        f.write(json.dumps(event, separators=(",", ":")) + "\n")
    print(f"appended {event['kind']} cycle={event['cycle_id']} → {path}", file=sys.stderr)
    return path


def read_payload(args: list[str]) -> dict:
    if args[:1] == ["--file"]:
        raw = Path(args[1]).read_text()
    elif args and args[0] != "-":
        raw = args[0]
    else:
        raw = sys.stdin.read()
    return json.loads(raw)


def main() -> None:
    load_env(ROOT / ".env")
    try:
        append(read_payload(sys.argv[1:]))
    except ContractError as err:
        raise SystemExit(str(err) or "invalid event") from err


if __name__ == "__main__":
    main()

File: tools/append-event.mjs
#!/usr/bin/env node
/**
 * Append one ledger event from a Grok bot.
 * Usage: node tools/append-event.mjs '{"kind":"quiet",...}'
 *        node tools/append-event.mjs --file ./payload.json
 */
import { spawnSync } from "node:child_process";
import { appendFileSync, readFileSync, mkdirSync, existsSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const root = resolve(dirname(fileURLToPath(import.meta.url)), "..");
loadEnv(join(root, ".env"));

const ledgerPath = resolve(process.env.LEDGER_PATH || join(root, "ledger", "events.jsonl"));

const raw = readPayload(process.argv.slice(2));
const event = typeof raw === "string" ? JSON.parse(raw) : raw;
checkContract(event);

mkdirSync(dirname(ledgerPath), { recursive: true });
appendFileSync(ledgerPath, JSON.stringify(event) + "\n", "utf8");
console.error(`appended ${event.kind} cycle=${event.cycle_id} → ${ledgerPath}`);

function checkContract(e) {
  const r = spawnSync(
    "python3",
    [join(root, "tools", "ledger_contract.py"), "--ledger", ledgerPath],
    { input: JSON.stringify(e), encoding: "utf8" }
  );
  if (r.status !== 0) {
    const msg = (r.stderr || r.stdout || "invalid event").trim();
    console.error(msg);
    process.exit(1);
  }
}

function readPayload(args) {
  if (args[0] === "--file") {
    return readFileSync(args[1], "utf8");
  }
  if (args[0] && args[0] !== "-") return args[0];
  return readFileSync(0, "utf8");
}

function loadEnv(path) {
  if (!existsSync(path)) return;
  for (const line of readFileSync(path, "utf8").split("\n")) {
    const t = line.trim();
    if (!t || t.startsWith("#")) continue;
    const i = t.indexOf("=");
    if (i < 1) continue;
    const key = t.slice(0, i).trim();
    const val = t.slice(i + 1).trim();
    if (!process.env[key]) process.env[key] = val;
  }
}
